File: .github/scripts/pr-analyzer.js
(async () => {
    const axios = require('axios');
    const { Octokit } = await import("@octokit/rest");
    const fs = require('fs');
    const path = require('path');
    const parseMidi = require("midi-file").parseMidi;
    const { Midi } = require("@tonejs/midi");
    
    const invalid_chars = [
        ":", "/", "\'", "\"", "?", "#", "%", "&", "{", "}", "\\", "<", ">", "*", "$",
        "!", "@", "+", "`", "|", "=", "."
    ];
    
    function filterFilename(name) {
        if (!name) {
            return ""
        }
        invalid_chars.forEach((c) => {
            name = name.split("").filter((i) => i !== c).join("");
        });
        return name;
    }
    
    function stringCompare(source, comparison) {
        // Returns a similarity score between two strings;
        let arr_a = source.toLowerCase().split("")
        let arr_b = comparison.toLowerCase().split("")
        const arr_a_copy = arr_a.slice()
        const arr_b_copy = arr_b.slice()
        let matching_characters_a = 0;
        let matching_characters_b = 0;
        // Check similarity for A->B
        arr_a.forEach(source_character => {
            const found_index = arr_b.indexOf(source_character)
            if (found_index > -1) {
                matching_characters_a += 1;
            }
            arr_b = arr_b.filter((item, index) => index != found_index);
        })
        // Check similarity for B->A
        arr_a = arr_a_copy.slice()
        arr_b = arr_b_copy.slice()
        arr_b.forEach(comparison_character => {
            const found_index = arr_a.indexOf(comparison_character)
            if (found_index > -1) {
                matching_characters_b += 1;
            }
            arr_a = arr_a.filter((item, index) => index != found_index);
        })
        // Determine score
        const a_score = matching_characters_a / source.length;
        const b_score = matching_characters_b / comparison.length;
        return (a_score + b_score) / 2;
    }
    
    function adjustRawURL(input_url) {
        return input_url.replace("github.com","raw.githubusercontent.com").replace("raw/","")
    }
    
    async function run() {
      try {
        class UploadHeader {
            constructor (name, mandatory) {
                this.name = name
                this.mandatory = mandatory
            }
        }
    
        // Response Flags
        let song_upload = false;
        let mandatory_headers_included = [];
        let unlisted_headers = [];
        let needs_changing = false;
        let new_game = false;
        let similar_game_name = "";
        let similar_game_score = 0;
        const number_vars = ["Tracks", "Duration"];
        const arr_vars = ["Tags"];
        const data_headers = [
            new UploadHeader("Game", true),
            new UploadHeader("Song", true),
            new UploadHeader("Category", true),
            new UploadHeader("Composers", false),
            new UploadHeader("Converters", false),
            new UploadHeader("Audio", false),
            new UploadHeader("Tags", false),
            new UploadHeader("Update Notes", false),
            new UploadHeader("Additional Notes", false),
            /* Automatically generated headers */
            // new UploadHeader("Date", true),
            // new UploadHeader("Verified", true),
            // new UploadHeader("Binary", true),
            // new UploadHeader("Duration", true),
            // new UploadHeader("Tracks", true),
        ]
    
    
        // PR Data
        const prNumber = process.env.PR_NUMBER;
        const repo = "theballaam96/candys-shop";
        const token = process.env.PAT_TOKEN;
        const response = await axios.get(`https://api.github.com/repos/${repo}/pulls/${prNumber}`, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
        const response_files = await axios.get(`https://api.github.com/repos/${repo}/pulls/${prNumber}/files`, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
        // Set file variables
        let bin_file = null;
        let midi_file = null;
        let midi_raw_file = null;
        let preview_file = null;
        let preview_extension = null;
        const preview_extensions = ["wav", "mp3"];
        for (i = 0; i < response_files.data.length; i++) {
          const f = response_files.data[i];
          if (f.filename) {
              const extension_sep = f.filename.split(".");
              const extension = extension_sep[extension_sep.length - 1];
              if (extension == "bin") {
                bin_file = f.filename;
              } else if (extension == "mid") {
                midi_file = f.filename;
                midi_raw_file = f.raw_url;
              } else if (preview_extensions.includes(extension)) {
                preview_file = f.filename;
                preview_extension = extension;
              }
          }
        }
        const octokit = new Octokit({auth: token})
    
        // Extract the PR message
        const prMessage = response.data.body;
        const rawPRData = prMessage ? prMessage.split("\r\n") : []
        const REQ_STRING = "IS SONG - DO NOT DELETE THIS LINE"
        if (rawPRData[0] == REQ_STRING) {
            song_upload = true;
        }
        let json_output = {}
        rawPRData.forEach((item, index) => {
            if ((index > 0) && (item)) {
                spl = item.split(/:(.*)/s)
                if (item.split("").includes(":")) {
                    key = spl[0].trim()
                    header_data = data_headers.filter(h => h.name == key)
                    if (header_data.length == 0) {
                        // Uses header not listed in the array
                        unlisted_headers.push(key)
                    } else {
                        if (header_data[0].mandatory) {
                            mandatory_headers_included.push(key)
                        }
                    }
                    json_output[key] = spl[1].trim()
                }
            }
        })
        number_vars.forEach(v => {
            if (Object.keys(json_output).includes(v)) {
                if (!isNaN(json_output[v])) {
                    json_output[v] = Number(json_output[v])
                }
            }
        })
        arr_vars.forEach(v => {
            if (Object.keys(json_output).includes(v)) {
                if (json_output[v]) {
                    json_output[v] = json_output[v].split(",").map(item => item.trim())
                }
            }
        })
        json_output["Verified"] = true;
        dt = new Date();
        json_output["Date"] = dt.toString();
    
        // Check headers
        let missing_mandatory_headers = [];
        data_headers.forEach(h => {
            if (h.mandatory) {
                if (!mandatory_headers_included.includes(h.name)) {
                    missing_mandatory_headers.push(h.name)
                }
            }
        })
        
    
        // Check game list
        const file = "images.json"
        const mappingFile = "mapping.json"
        const filePath = path.join(__dirname, `../../${file}`);
        const mappingPath = path.join(__dirname, `../../${mappingFile}`);
        const imageData = fs.existsSync(filePath) ? require(filePath) : [];
        const mappingData = fs.existsSync(mappingPath) ? require(mappingPath) : [];
        // Get new file name
        let revisions = mappingData.filter((entry) => ((entry["Game"] == json_output["Game"]) && (entry["Song"] == json_output["Song"]))).length;
        const rev_string = revisions == 0 ? "" : ` (REV ${revisions})`;
        const sub_file = `${filterFilename(json_output["Game"])}/${filterFilename(json_output["Song"])}${rev_string}`
        if (preview_file) {
          json_output["Audio"] = encodeURI(`https://github.com/theballaam96/candys-shop/raw/main/previews/${sub_file}.${preview_extension}`)
        }
        if (bin_file) {
          json_output["Binary"] = `binaries/${sub_file}.bin`
        }
        if (midi_file) {
            const midiURL = adjustRawURL(midi_raw_file);
            const midiParsed = await Midi.fromUrl(midiURL);
            if (midiParsed.duration) {
                json_output["Duration"] = midiParsed.duration;
            }
        }
    
        if (song_upload) {
            needs_changing = missing_mandatory_headers.length > 0 || unlisted_headers.length > 0 || !Object.keys(json_output).includes("Binary") || !Object.keys(json_output).includes("Audio");
        }
    
        const game_name = Object.keys(json_output).includes("Game") ? json_output["Game"] : null;
        const similarity_threshold = 0.75
        if (game_name != null) {
            if (!Object.keys(imageData).includes(game_name)) {
                new_game = true;
                let max_score = 0;
                let max_score_name = "";
                Object.keys(imageData).forEach(gn => {
                    new_score = stringCompare(game_name, gn);
                    if (new_score > max_score) {
                        max_score = new_score
                        max_score_name = gn
                    }
                })
                if (max_score >= similarity_threshold) {
                    similar_game_name = max_score_name;
                    similar_game_score = max_score;
                }
            }
        }
    
        // Compose Message
        let segments = [
            "Mornin'",
            "",
            "I've analyzed your pull request and ascertained the following information from it. This will help the verifiers handle your request faster:",
            `> Is Song Upload: ${song_upload ? "Yes": "No"}`
        ]
        if (song_upload) {
            segments.push(`> Has Binary File: ${Object.keys(json_output).includes("Binary") ? "Yes" : "No"}`)
            segments.push(`> Has Preview: ${Object.keys(json_output).includes("Audio") ? "Yes" : "No"}`)
            segments.push(`> Missing Mandatory Information: ${missing_mandatory_headers.length == 0 ? "None" : missing_mandatory_headers.join(", ")}`)
            segments.push(`> Headers which I don't understand: ${unlisted_headers.length == 0 ? "None": unlisted_headers.join(", ")}`)
            segments.push(`> Is new game: ${new_game ? "Yes": "No"}`)
            if (new_game) {
                if (similar_game_score > 0) {
                    segments.push(`> Detected as Similar to: ${similar_game_name} (Similarity Score: ${parseInt(similar_game_score * 100, 10)}%)`)
                }
            }
        }
        segments.push(`> Something needs changing: ${needs_changing ? "YES!!! (PLEASE ENSURE YOU FIX ANY ERRORS SO THIS SAYS NO BEFORE MERGING)": "No"}`)
        segments.push("") // Prevent following messges getting indented
        if (song_upload) {
            segments.push("Here's what the output will look like:")
            segments.push("\`\`\`")
            segments.push(JSON.stringify(json_output, undefined, 4))
            segments.push("\`\`\`")
        }
    
    
        const message = segments.join("\n");
        await octokit.issues.createComment({
            owner: repo.split("/")[0],
            repo: repo.split("/")[1],
            issue_number: parseInt(prNumber, 10), // Ensure prNumber is parsed as an integer
            body: message,
        });
    
        if (needs_changing) {
            console.error("Something needs changing in order to make this PR valid.")
            process.exit(1);
        }
      } catch (error) {
        console.error('Error:', error.response ? error.response.data : error.message || error);
        process.exit(1);
      }
    }
    
    run();
})();

File: .github/scripts/python/pr_analyzer.py
import os
import sys
import json
import requests
from datetime import datetime
from github import Github, Auth
from dotenv import load_dotenv
from donk_lib import pull_pr_data, getPath, postAudio, adjust_raw_url

load_dotenv()

POST_COMMENT = True

def message(pr_data):
    webhook_url = os.getenv("DISCORD_WEBHOOK_SUBMISSIONS")
    triggered_action = os.getenv("TRIGGERED_ACTION")
    header_text = {
        "closed": {
            "text": "A pull request was closed",
            "color": 0xFF0000,
        },
        "ready_for_review": {
            "text": "A pull request was marked as ready for review",
            "color": 0xD700A7,
        },
        "reopened": {
            "text": "A pull request was reopened",
            "color": 0x9FD700,
        },
        "edited": {
            "text": "A pull request was edited",
            "color": 0xFCCA03,
        },
        "synchronize": {
            "text": "A pull request was synchronized",
            "color": 0xFCCA03,
        },
        "opened": {
            "text": "New Pull Request",
            "color": 0x03FC0F,
        },
    }
    default_text = f"A pull request action ({triggered_action})"
    title = f"{header_text.get(triggered_action, {'text': default_text})['text']} from {pr_data['submitter']}"
    components = [
        {
            "type": 1,  # Action row
            "components": [],
        }
    ]
    has_component = False
    has_embed = False
    audio_file = None
    if pr_data["is_song"]:
        json_output = pr_data["output"]
        color = header_text.get(triggered_action, {'color': 0x03FC0F})['color']
        # Set up description
        desc_information = {
            "Game": json_output.get("Game", "Not Provided"),
            "Song Name": json_output.get("Song", "Not Provided"),
            "Original Composer": json_output.get("Composers", "Not Provided"),
            "Converted By": json_output.get("Converters", "Not Provided"),
            "Type": json_output.get("Category", "Not Provided"),
            "Tags": ", ".join(json_output.get("Tags", ["Not Provided"])),
            "Needs a logo": pr_data.get("new_game", False),
            "Duration": json_output.get("Duration", 0),
            "Update Notes": json_output.get("Update Notes", "Not Provided"),
            "Additional Notes": json_output.get("Additional Notes", "Not Provided"),
        }

        desc_strings = [f"**{k}**: {v}" for k, v in desc_information.items()]
        desc = '\n'.join(desc_strings)
        # Set up embeds
        embed = {
            "title": "New Song Pull Request",
            "color": color,
            "description": desc,
            "timestamp": datetime.utcnow().isoformat(),
        }

        buttons = {
            "Pull Request": pr_data.get("pull_request"),
            "Binary File": pr_data.get("bin_raw"),
            "MIDI File": pr_data.get("midi_raw"),
            "YouTube": json_output.get("Audio"),
        }
        if pr_data.get("preview_raw") is not None:
            url_to_pull = adjust_raw_url(pr_data["preview_raw"])
            audio_resp = requests.get(url_to_pull)
            audio_resp.raise_for_status()
            audio_file = audio_resp.content
        for but_title, prop in buttons.items():
            if prop is None:
                continue
            components[0]["components"].append({
                "type": 2,  # Button
                "style": 5,  # Link button
                "label": but_title,
                "url": prop,
            })
        has_component = True
        has_embed = True
    else:
        components[0]["components"].append({
            "type": 2,  # Button
            "style": 5,  # Link button
            "label": "Pull Request",
            "url": pr_data["pull_request"],
        })
        has_component = True

    payload = {
        "content": title,
    }
    if has_embed:
        payload["embeds"] = [embed]
    if has_component:
        webhook_url = f"{webhook_url}?wait=true&with_components=true"
        payload["components"] = components

    # Post the embed
    try:
        whresp = requests.post(webhook_url, json=payload, headers={"Content-Type": "application/json"})
        whresp.raise_for_status()
        print("Message sent successfully:", whresp.text)
    except Exception as e:
        print("Error sending webhook embed:", e)
        sys.exit(1)
    if audio_file is not None:
        json_output = pr_data["output"]
        postAudio(json_output.get("Song", "Not Provided"), pr_data["preview_ext"], audio_file, webhook_url)
    return

def run():
    try:
        needs_changing = False
        new_game = False
        pr_number = os.getenv("PR_NUMBER")
        token = os.getenv("PAT_TOKEN")
        repo_full = "theballaam96/candys-shop"
                
        data = pull_pr_data(pr_number)
        song_upload = data["is_song"]
        if song_upload:
            json_output = data["output"]
            print(json_output)

            # Load image/mapping JSONs (optional)
            file_path = getPath("images.json")
            image_data = json.load(open(file_path)) if os.path.exists(file_path) else {}

            # Placeholder for MIDI duration (skip actual parsing)

            if song_upload:
                needs_changing = (not data.get("bin")) or ((not json_output.get("Audio")) and (not data.get("preview")))

            # Check for new game similarity
            game_name = json_output.get("Game")
            if game_name and isinstance(image_data, dict):
                if game_name not in image_data:
                    new_game = True

        # Compose message
        segments = [
            "Mornin'",
            "",
            "I've analyzed your pull request and ascertained the following information from it. This will help the verifiers handle your request faster:",
            f"> Is Song Upload: {'Yes' if song_upload else 'No'}",
        ]
        if song_upload:
            segments.append(f"> Has Binary File: {'Yes' if 'Binary' in json_output else 'No'}")
            segments.append(f"> Has Preview: {'Yes' if 'Audio' in json_output else 'No'}")
            segments.append(f"> Is new game: {'Yes' if new_game else 'No'}")
        segments.append(f"> Something needs changing: {'YES!!! (PLEASE FIX ERRORS)' if needs_changing else 'No'}")
        if song_upload:
            segments.append("")
            segments.append("Here's what the output will look like:")
            segments.append("```")
            segments.append(json.dumps(json_output, indent=4))
            segments.append("```")

        prMessage = "\n".join(segments)

        # Comment on PR
        if POST_COMMENT:
            auth = Auth.Token(token)
            g = Github(auth=auth)
            repo = g.get_repo(repo_full)
            issue = repo.get_issue(int(pr_number))
            issue.create_comment(prMessage)
        else:
            print(prMessage)

        if needs_changing:
            print("::error::Something needs changing in order to make this PR valid.")
            exit(1)

        message(data)

    except Exception as e:
        print(f"::error::{str(e)}")
        exit(1)


if __name__ == "__main__":
    run()
